aten/src/ATen/native/Gesv.h

```diff
@@ -1,30 +1,94 @@
+#include <utility>
 #include "ATen/ATen.h"
 
 namespace at { namespace native {
 
-static inline void checkInputs(const Tensor& self, const Tensor& A) {
-  if (A.size(-1) != A.size(-2)) {
-    AT_ERROR("A must be batches of square matrices, "
-        "but they are %lld by %lld matrices",
-        (long long)A.size(-1), (long long)A.size(-2));
+static inline bool isTransposeContiguous(Tensor& self) {
+  return self.dim() == 2 &&
+          self.stride(0) == 1 &&
+          self.stride(1) == self.size(0);
+}
+
+/* gesv takes (self, A) and returns (sol, lu).
+ * (i)  output tensors (sol, lu) may be same as input tensors (self, A)
+ * (ii) for 2D matrices, .t_() represents their column-major format
+ *
+ * Before passing pointers to Lapack, we need to ensure that these pointers
+ * represent Fortran-contiguous tensors in column-major format
+ *
+ * Cases:
+ * 1) `out` has correct shape but elements do not form a contiguous
+ * chunk of memory. Since shape is correct, we don't resize_ it. Instead, we
+ * clone the input tensor into a buffer, use the buffer for Lapack and finally
+ * copy the buffer to the output tensor.
+ *
+ * 2) out.t() is contiguous:
+ *    (i)  &in == &out: use out.data() as is. Do nothing
```

Comment:
This is a little overengineered; we don't guarantee that passing in the input tensor as the output tensor works for any other operation in pytorch. Since your code is correct that is nice to have, but I'm not sure a user will stumble upon this.

Reply:
The main reason for these optimizations is that the Lapack `gesv` interface takes input tensors which contain the outputs after the function is called, so it's nice to support that. Also, `test_gesv` explicitly checks for this case.